### DISCUSSION TITLE
Added: LDM Notes for Nov. 18, 2019

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#3032] [General]
https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-11-18.md

Agenda: discussion of more pattern-matching features

**Comment by Joe4evr** [comment]
> The proposal contains three breaking changes:

But there's only two points listed. 🤔

**Comment by VBAndCs** [comment]
As I said before, I am not a fan of using `and` and `or` unless they are also allowed in if conditions!
Use `&` and `|` instead, to maintain the same rules all over the syntax.

**Comment by HaloFour** [comment]
@VBAndCs 

The entire point of adding `and` and `or` is to disambiguate the patterns from the Boolean operands where they would be ambiguous.  If they could be used interchangeably that would completely defeat the purpose.

```
if (obj is bool && false) { ... }
if (obj is bool and false) { ... }
```

**Comment by VBAndCs** [comment]
We are already familiar with using  `()` in conditions for disambiguity, so we can do the same with patterns.

**Comment by HaloFour** [comment]
@VBAndCs 

Parenthesis are for establishing order of operation, not disambiguation.  And parenthetical patterns are also proposed.  Patterns behave differently from operators, trying to combine their syntax, even without the ambiguity, would only be confusing.

**Comment by DavidArno** [comment]
I *really* don't like the idea that these two would behave differently due to parentheses:
```cs
if (obj is bool && false) // always false 
if (obj is (bool && false)) // true if obj is false
```
I'm 100% with @HaloFour here, the following looks like a pattern rather than a logical expression and so makes it clear how the `false` is being used:
``` cs
if (obj is bool and false)
```

**Comment by FiniteReality** [comment]
I'm not sure if this has been suggested already, but for the binary operators issue, can't we do something like this?

```cs
object o = (short)1;
_ = o switch
{
    int x < 3 => 0,
    _ == 1 => 0xC0FFEE,
    _ => 0xDEADBEEF
}
```

That way we're explicitly including the type (`int x`) and also satisfying not "using a binary operator as a unary operator" (`x < 3`).
This also (potentially) solves the type issue, as we could use `var x < 3`... However, that seems kind-of strange, and I can't think of a good way to represent that in current code.
